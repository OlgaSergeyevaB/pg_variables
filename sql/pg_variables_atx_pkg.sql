--
-- PGPRO-7614: function pgv_free() inside autonomous transaction
--
select pgv_free();
--
--
-- Functions pgv_free() + pgv_get() inside autonomous transaction; package
-- with regular variable; autonomous transaction with commit.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
	BEGIN AUTONOMOUS;
		SELECT pgv_get('vars', 'int1', null::int);
		SELECT pgv_free();
-- ERROR:  unrecognized package "vars"
		SELECT pgv_get('vars', 'int1', null::int);
	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- regular variable; autonomous transaction with commit.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
	BEGIN AUTONOMOUS;
		SELECT pgv_free();
	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- regular variable; autonomous transaction with rollback.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
	BEGIN AUTONOMOUS;
		SELECT pgv_free();
	ROLLBACK;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- transactional variable; autonomous transaction with rollback.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1, true);
	BEGIN AUTONOMOUS;
		SELECT pgv_free();
	ROLLBACK;
	SELECT pgv_get('vars', 'int1', null::int);
ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- transactional variable; autonomous transaction with commit.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1, true);
	BEGIN AUTONOMOUS;
		SELECT pgv_free();
	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ROLLBACK;
--
--
-- Function pgv_free() inside recursive autonomous transactions.
--
BEGIN;
	BEGIN AUTONOMOUS;
		SELECT pgv_set('vars', 'int1', 1);
		BEGIN AUTONOMOUS;
			BEGIN AUTONOMOUS;
				SELECT pgv_free();
			COMMIT;
-- ERROR:  unrecognized package "vars"
			SELECT pgv_get('vars', 'int1', null::int);
		COMMIT;
-- ERROR:  unrecognized package "vars"
		SELECT pgv_get('vars', 'int1', null::int);
	COMMIT;
ROLLBACK;
--
--
-- Function pgv_free() inside recursive autonomous transactions;
-- recreating the package after deletion with using regular
-- variable.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
	BEGIN AUTONOMOUS;
		BEGIN AUTONOMOUS;
			SELECT pgv_get('vars', 'int1', null::int);
			BEGIN AUTONOMOUS;
				SELECT pgv_free();
			COMMIT;
-- ERROR:  unrecognized package "vars"
			SELECT pgv_get('vars', 'int1', null::int);
		COMMIT;
		SELECT pgv_set('vars', 'int1', 2);
	COMMIT;
	SELECT pgv_get('vars', 'int1', null::int);
ROLLBACK;
--
--
-- Function pgv_free() inside recursive autonomous transactions;
-- recreating the package after deletion with using transactional
-- variable.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
	BEGIN AUTONOMOUS;
		BEGIN AUTONOMOUS;
			SELECT pgv_get('vars', 'int1', null::int);
			BEGIN AUTONOMOUS;
				SELECT pgv_free();
			COMMIT;
-- ERROR:  unrecognized package "vars"
			SELECT pgv_get('vars', 'int1', null::int);
		COMMIT;
		SELECT pgv_set('vars', 'int1', 2, true);
		SELECT pgv_list();
	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ROLLBACK;
--
--
-- Do not free hash_seq_search scans of parent transaction. 
--
BEGIN;
	SELECT pgv_insert('test', 'x', row (1::int, 2::int), false);
	DECLARE r1_cur CURSOR FOR SELECT pgv_select('test', 'x');
	FETCH 1 IN r1_cur;
	BEGIN AUTONOMOUS;
	ROLLBACK;
ROLLBACK;

select pgv_free();
