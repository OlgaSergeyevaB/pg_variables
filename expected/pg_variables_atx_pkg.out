--
-- PGPRO-7614: function pgv_free() inside autonomous transaction
--
select pgv_free();
 pgv_free 
----------
 
(1 row)

--
--
-- Functions pgv_free() + pgv_get() inside autonomous transaction; package
-- with regular variable; autonomous transaction with commit.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
 pgv_set 
---------
 
(1 row)

	BEGIN AUTONOMOUS;
		SELECT pgv_get('vars', 'int1', null::int);
 pgv_get 
---------
       1
(1 row)

		SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

-- ERROR:  unrecognized package "vars"
		SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- regular variable; autonomous transaction with commit.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
 pgv_set 
---------
 
(1 row)

	BEGIN AUTONOMOUS;
		SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- regular variable; autonomous transaction with rollback.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
 pgv_set 
---------
 
(1 row)

	BEGIN AUTONOMOUS;
		SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

	ROLLBACK;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- transactional variable; autonomous transaction with rollback.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1, true);
 pgv_set 
---------
 
(1 row)

	BEGIN AUTONOMOUS;
		SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

	ROLLBACK;
	SELECT pgv_get('vars', 'int1', null::int);
 pgv_get 
---------
       1
(1 row)

ROLLBACK;
--
--
-- Function pgv_free() inside autonomous transaction; package with
-- transactional variable; autonomous transaction with commit.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1, true);
 pgv_set 
---------
 
(1 row)

	BEGIN AUTONOMOUS;
		SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
ROLLBACK;
--
--
-- Function pgv_free() inside recursive autonomous transactions.
--
BEGIN;
	BEGIN AUTONOMOUS;
		SELECT pgv_set('vars', 'int1', 1);
 pgv_set 
---------
 
(1 row)

		BEGIN AUTONOMOUS;
			BEGIN AUTONOMOUS;
				SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

			COMMIT;
-- ERROR:  unrecognized package "vars"
			SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
		COMMIT;
-- ERROR:  unrecognized package "vars"
		SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
	COMMIT;
ROLLBACK;
--
--
-- Function pgv_free() inside recursive autonomous transactions;
-- recreating the package after deletion with using regular
-- variable.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
 pgv_set 
---------
 
(1 row)

	BEGIN AUTONOMOUS;
		BEGIN AUTONOMOUS;
			SELECT pgv_get('vars', 'int1', null::int);
 pgv_get 
---------
       1
(1 row)

			BEGIN AUTONOMOUS;
				SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

			COMMIT;
-- ERROR:  unrecognized package "vars"
			SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
		COMMIT;
		SELECT pgv_set('vars', 'int1', 2);
 pgv_set 
---------
 
(1 row)

	COMMIT;
	SELECT pgv_get('vars', 'int1', null::int);
 pgv_get 
---------
       2
(1 row)

ROLLBACK;
--
--
-- Function pgv_free() inside recursive autonomous transactions;
-- recreating the package after deletion with using transactional
-- variable.
--
BEGIN;
	SELECT pgv_set('vars', 'int1', 1);
 pgv_set 
---------
 
(1 row)

	BEGIN AUTONOMOUS;
		BEGIN AUTONOMOUS;
			SELECT pgv_get('vars', 'int1', null::int);
 pgv_get 
---------
       1
(1 row)

			BEGIN AUTONOMOUS;
				SELECT pgv_free();
 pgv_free 
----------
 
(1 row)

			COMMIT;
-- ERROR:  unrecognized package "vars"
			SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
		COMMIT;
		SELECT pgv_set('vars', 'int1', 2, true);
 pgv_set 
---------
 
(1 row)

		SELECT pgv_list();
   pgv_list    
---------------
 (vars,int1,t)
(1 row)

	COMMIT;
-- ERROR:  unrecognized package "vars"
	SELECT pgv_get('vars', 'int1', null::int);
ERROR:  unrecognized package "vars"
ROLLBACK;
--
--
-- Do not free hash_seq_search scans of parent transaction. 
--
BEGIN;
	SELECT pgv_insert('test', 'x', row (1::int, 2::int), false);
 pgv_insert 
------------
 
(1 row)

	DECLARE r1_cur CURSOR FOR SELECT pgv_select('test', 'x');
	FETCH 1 IN r1_cur;
 pgv_select 
------------
 (1,2)
(1 row)

	BEGIN AUTONOMOUS;
	ROLLBACK;
ROLLBACK;
select pgv_free();
 pgv_free 
----------
 
(1 row)

